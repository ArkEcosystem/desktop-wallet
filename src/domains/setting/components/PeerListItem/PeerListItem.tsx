import { Circle } from "app/components/Circle";
import { Dropdown } from "app/components/Dropdown";
import { Icon } from "app/components/Icon";
import React from "react";

type PeerListItemProps = {
	coin?: string;
	coinClass?: string;
	name?: string;
	peerIp?: string;
	type?: string;
	actions?: any[];
	onAction?: any;
};

export const PeerListItem = ({ coin, coinClass, name, peerIp, type, actions, onAction }: PeerListItemProps) => (
	<tr className="border-b border-theme-neutral-200">
		<td className="inline-flex items-center py-4 mt-1">
			<Circle className={coinClass}>
				<Icon name={coin!} />
			</Circle>
			<span className="ml-2">{coin?.toLocaleUpperCase()}</span>
		</td>
		<td className="py-1">
			<span>{name}</span>
		</td>
		<td className="py-1">
			<span>{peerIp}</span>
		</td>
		<td className="py-1 text-theme-primary-300">
			<div className="flex justify-center h-full">
				<Icon name={type!} />
			</div>
		</td>
		<td>
			{actions?.length &&
				(() => {
					return <Dropdown options={actions} onSelect={onAction} />;
				})()}
		</td>
	</tr>
);

PeerListItem.defaultProps = {
	actions: [],
};
