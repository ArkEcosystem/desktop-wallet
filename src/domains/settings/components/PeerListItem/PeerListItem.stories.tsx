import { Table } from "app/components/Table";
import React from "react";

import { PeerListItem } from "./PeerListItem";

export default { title: "Settings / Components / PeerListItem" };

const data = [
	{
		coin: "Btc",
		coinClass: "text-theme-warning-400 border-theme-warning-200",
		name: "OLEBank",
		peerIp: "194.168.4.67:800",
		type: "Multisig",
		actions: [
			{
				label: "Edit",
				value: "edit",
			},
			{
				label: "Delete",
				value: "delete",
			},
		],
	},
];

const columns = [
	{
		Header: "Network",
	},
	{
		Header: "Name",
	},
	{
		Header: "Peer IP",
	},
	{
		Header: "Type",
		className: "flex justify-center",
	},
];

export const Default = () => (
	<Table columns={columns} data={data}>
		{(rowData: any) => <PeerListItem {...rowData} />}
	</Table>
);
