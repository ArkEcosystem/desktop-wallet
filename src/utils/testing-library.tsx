import { ARK } from "@arkecosystem/platform-sdk-ark";
import { Environment } from "@arkecosystem/platform-sdk-profiles";
import { render } from "@testing-library/react";
import { EnvironmentProvider } from "app/contexts";
import { i18n } from "app/i18n";
import { httpClient } from "app/services";
import { createMemoryHistory } from "history";
import React from "react";
import { I18nextProvider } from "react-i18next";
import { Router } from "react-router-dom";
import { StubStorage } from "tests/mocks";

import envFixture from "../tests/fixtures/env/data.json";

export const env = new Environment({ coins: { ARK }, httpClient, storage: new StubStorage(envFixture) });

const WithProviders: React.FC = ({ children }: { children?: React.ReactNode }) => (
	<I18nextProvider i18n={i18n}>
		<EnvironmentProvider env={env}>{children}</EnvironmentProvider>
	</I18nextProvider>
);

const customRender = (component: React.ReactElement, options: any = {}) =>
	render(component, { wrapper: WithProviders, ...options });

const renderWithRouter = (
	component: React.ReactElement,
	{ routes = ["/"], history = createMemoryHistory({ initialEntries: routes }), withProviders = true } = {},
) => {
	const RouterWrapper = ({ children }: { children: React.ReactNode }) =>
		withProviders ? (
			<WithProviders>
				<Router history={history}>{children}</Router>
			</WithProviders>
		) : (
			<Router history={history}>{children}</Router>
		);

	return {
		...customRender(component, { wrapper: RouterWrapper }),
		history,
	};
};

export * from "@testing-library/react";

export { customRender as render, renderWithRouter };
