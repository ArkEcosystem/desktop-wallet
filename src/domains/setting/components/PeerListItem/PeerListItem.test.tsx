import React from "react";
import { act, fireEvent, render } from "testing-library";

import { PeerListItem } from "./PeerListItem";

const peer = {
	coin: "ark",
	network: "devnet",
	name: "ROBank",
	host: "http://167.114.29.48:4003/api",
	isMultiSignature: false,
};

describe("PeerListItem", () => {
	it("should render", () => {
		const { container } = render(
			<table>
				<tbody>
					<PeerListItem {...peer} isMultiSignature={true} />
				</tbody>
			</table>,
		);

		expect(container).toMatchSnapshot();
	});

	it("should trigger onAction callback if provided", () => {
		const onAction = jest.fn();
		const options = [
			{ label: "Option 1", value: "1" },
			{ label: "Option 2", value: "2" },
		];

		const { container, getByTestId } = render(
			<table>
				<tbody>
					<PeerListItem {...peer} options={options} onAction={onAction} />
				</tbody>
			</table>,
		);
		const toggle = getByTestId("dropdown__toggle");

		act(() => {
			fireEvent.click(toggle);
		});

		expect(getByTestId("dropdown__content")).toBeTruthy();

		const firstOption = getByTestId("dropdown__option--0");
		expect(firstOption).toBeTruthy();

		act(() => {
			fireEvent.click(firstOption);
		});

		expect(container.querySelectorAll("ul").length).toEqual(0);
		expect(onAction).toHaveBeenCalled();
	});

	it("should ignore onAction callback if not provided", () => {
		const options = [
			{ label: "Option 1", value: "1" },
			{ label: "Option 2", value: "2" },
		];

		const { container, getByTestId } = render(
			<table>
				<tbody>
					<PeerListItem {...peer} options={options} />
				</tbody>
			</table>,
		);
		const toggle = getByTestId("dropdown__toggle");

		act(() => {
			fireEvent.click(toggle);
		});

		expect(getByTestId("dropdown__content")).toBeTruthy();

		const firstOption = getByTestId("dropdown__option--0");
		expect(firstOption).toBeTruthy();

		act(() => {
			fireEvent.click(firstOption);
		});

		expect(container.querySelectorAll("ul").length).toEqual(0);
	});
});
