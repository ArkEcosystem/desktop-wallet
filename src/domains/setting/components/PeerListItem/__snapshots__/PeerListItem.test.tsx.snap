// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PeerListItem should render 1`] = `
<div>
  <table>
    <tbody>
      <tr
        class="border-b border-theme-neutral-200"
      >
        <td
          class="inline-flex items-center py-4 mt-1"
        >
          <div
            class="sc-AxjAm bLfakz"
          >
            <div
              class="sc-AxirZ iQaTP"
              height="1em"
              width="1em"
            >
              <svg>
                btc.svg
              </svg>
            </div>
          </div>
          <span
            class="ml-2"
          >
            BITCOIN
          </span>
        </td>
        <td
          class="py-1"
        >
          <span />
        </td>
        <td
          class="py-1"
        >
          <span />
        </td>
        <td
          class="py-1 text-theme-neutral"
        >
          <div
            class="flex justify-center h-full"
          >
            <div
              class="sc-AxirZ iQaTP"
              height="1em"
              width="1em"
            />
          </div>
        </td>
        <td
          class="text-theme-neutral-300"
        >
          0
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
